File: src/data-parser/CSVParser.js
/**
 * CSVParser - CSV 파일 파싱
 * 주석 처리, 큰따옴표 이스케이프, 타입 자동 변환 지원
 */
import { gameLogger } from '../core/Logger.js';

class CSVParser {
    /**
     * CSV 텍스트 파싱
     * @param {string} text - CSV 텍스트
     * @returns {Object[]} 객체 배열
     */
    static parse(text) {
        const lines = text.split(/\r?\n/);
        const result = [];
        
        // 헤더 행 추출
        let headerIndex = 0;
        while (headerIndex < lines.length) {
            const line = lines[headerIndex].trim();
            // 주석이나 빈 행 스킵
            if (line && !line.startsWith('#')) {
                break;
            }
            headerIndex++;
        }
        
        if (headerIndex >= lines.length) {
            gameLogger.warn('CSV has no data');
            return [];
        }
        
        const headers = this.parseLine(lines[headerIndex]);
        
        // 데이터 행 파싱
        for (let i = headerIndex + 1; i < lines.length; i++) {
            const line = lines[i].trim();
            
            // 주석, 빈 행 스킵
            if (!line || line.startsWith('#')) continue;
            
            const values = this.parseLine(line);
            const row = {};
            
            headers.forEach((header, index) => {
                const value = values[index];
                row[header] = this.convertType(value);
            });
            
            result.push(row);
        }
        
        return result;
    }

    /**
     * CSV 행 파싱 (큰따옴표 처리 포함)
     * @param {string} line 
     * @returns {string[]}
     */
    static parseLine(line) {
        const result = [];
        let current = '';
        let inQuotes = false;
        
        for (let i = 0; i < line.length; i++) {
            const char = line[i];
            
            if (char === '"') {
                // 큰따옴표 이스케이프 확인
                if (inQuotes && line[i + 1] === '"') {
                    current += '"';
                    i++; // 다음 문자 스킵
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (char === ',' && !inQuotes) {
                result.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }
        
        // 마지막 필드 추가
        result.push(current.trim());
        
        return result;
    }

    /**
     * 타입 자동 변환
     * @param {string} value 
     * @returns {number|boolean|string|Object}
     */
    static convertType(value) {
        // 빈 문자열
        if (value === '') return null;
        
        // 불리언
        if (value.toLowerCase() === 'true') return true;
        if (value.toLowerCase() === 'false') return false;
        
        // 숫자
        if (/^-?\d+(\.\d+)?$/.test(value)) {
            return parseFloat(value);
        }
        
        // JSON 문자열 (중괄호로 시작)
        if (value.startsWith('{') && value.endsWith('}')) {
            try {
                // 큰따옴표 복원 (CSV 에서 이스케이프된 것)
                const jsonStr = value.replace(/""/g, '"');
                return JSON.parse(jsonStr);
            } catch (e) {
                gameLogger.warn('Failed to parse JSON value:', value);
                return value;
            }
        }
        
        // 문자열
        return value;
    }

    /**
     * CSV 파일에서 로드 (fetch)
     * @param {string} url - CSV 파일 URL
     * @returns {Promise<Object[]>}
     */
    static async parseFile(url) {
        try {
            const response = await fetch(url);
            if (!response.ok) {
                throw new Error(`Failed to fetch ${url}: ${response.status}`);
            }
            const text = await response.text();
            return this.parse(text);
        } catch (error) {
            gameLogger.error(`Failed to load CSV ${url}:`, error);
            throw error;
        }
    }
}

export { CSVParser };

File: assets/Scripts/Core/CSVParser.cs
using System;
using System.Collections.Generic;

/// <summary>
/// CSV 파일 파서
/// 주석 처리, 큰따옴표 이스케이프, 타입 자동 변환 지원
/// </summary>
public static class CSVParser
{
    /// <summary>
    /// CSV 텍스트 파싱
    /// </summary>
    public static List<Dictionary<string, object>> Parse(string text)
    {
        var result = new List<Dictionary<string, object>>();
        if (string.IsNullOrEmpty(text)) return result;

        var lines = text.Split(new[] { "\r\n", "\n" }, StringSplitOptions.RemoveEmptyEntries);
        
        // 헤더 행 추출 (주석/빈 행 스킵)
        int headerIndex = 0;
        while (headerIndex < lines.Length)
        {
            var line = lines[headerIndex].Trim();
            if (!string.IsNullOrEmpty(line) && !line.StartsWith("#"))
                break;
            headerIndex++;
        }
        
        if (headerIndex >= lines.Length) return result;
        
        var headers = ParseLine(lines[headerIndex]);
        
        // 데이터 행 파싱
        for (int i = headerIndex + 1; i < lines.Length; i++)
        {
            var line = lines[i].Trim();
            if (string.IsNullOrEmpty(line) || line.StartsWith("#")) continue;
            
            var values = ParseLine(line);
            var row = new Dictionary<string, object>();
            
            for (int j = 0; j < headers.Count && j < values.Count; j++)
            {
                row[headers[j]] = ConvertType(values[j]);
            }
            
            result.Add(row);
        }
        
        return result;
    }

    /// <summary>
    /// CSV 행 파싱 (큰따옴표 처리 포함)
    /// </summary>
    private static List<string> ParseLine(string line)
    {
        var result = new List<string>();
        var current = "";
        var inQuotes = false;
        
        for (int i = 0; i < line.Length; i++)
        {
            var c = line[i];
            
            if (c == '"')
            {
                if (inQuotes && i + 1 < line.Length && line[i + 1] == '"')
                {
                    current += '"';
                    i++;
                }
                else
                {
                    inQuotes = !inQuotes;
                }
            }
            else if (c == ',' && !inQuotes)
            {
                result.Add(current.Trim());
                current = "";
            }
            else
            {
                current += c;
            }
        }
        
        result.Add(current.Trim());
        return result;
    }

    /// <summary>
    /// 타입 자동 변환
    /// </summary>
    private static object ConvertType(string value)
    {
        if (int.TryParse(value, out var intValue)) return intValue;
        if (float.TryParse(value, out var floatValue)) return floatValue;
        if (bool.TryParse(value, out var boolValue)) return boolValue;
        return value;
    }
}
